l5kit/l5kit/simulation/dataset.py

```diff
@@ -0,0 +1,217 @@
+from copy import deepcopy
+from typing import Dict, List, Set, Tuple
+
+import numpy as np
+from torch.utils.data import Dataset
+
+from l5kit.data import filter_agents_by_frames, PERCEPTION_LABEL_TO_INDEX
+from l5kit.dataset import EgoDataset
+from l5kit.geometry.transform import yaw_as_rotation33
+from l5kit.simulation.utils import disable_agents, insert_agent
+
+
+class SimulationDataset(Dataset):
+    def __init__(self, dataset: EgoDataset, scene_indices: List[int], start_frame: int = 0,
+                 disable_new_agents: bool = False, distance_th_far: float = 15,
+                 distance_th_close: float = 10) -> None:
+        """This class allows to:
+        - rasterise the same frame across multiple scenes for ego;
+        - rasterise the same frame across multiple scenes for multiple agents;
+        - filter agents based on distance to ego;
+        - set ego in future frames;
+        - set agents in future frames;
+
+        .. note:: only vehicles (car label) are picked as agents
+
+        :param dataset: the original ego dataset
+        :param scene_indices: the indices of the scenes to take
+        :param start_frame: the unroll start frame, use only if disable_new_agents is True
+        :param disable_new_agents: if to disable new agents coming in after start_frame, defaults to False
+        :param distance_th_far: the distance threshold for new agents,
+            An agent will be grabbed if it's closer than this value to ego, defaults to 15
+        :param distance_th_close: the distance threshold for already tracked agents,
+            An agent will be grabbed if it's closer than this value to ego, defaults to 10
+        """
+        self.scene_indices = scene_indices
+        self.filter_agents_thr = dataset.cfg["raster_params"]["filter_agents_threshold"]
+
+        self.scene_dataset_batch: Dict[int, EgoDataset] = {}  # dicts preserve insertion order
+        for scene_idx in self.scene_indices:
+            scene_dataset = dataset.get_scene_dataset(scene_idx)
+            self.scene_dataset_batch[scene_idx] = scene_dataset
```

Comment:
I think it would be better to have the `SimulationDataset` to accept a parameter of the type `scene_dataset_batch: Dict[int, EgoDataset]` and this loop is done by the caller of this class.

Reply:
I'm not sure about this point, as I would need to add a function to perform the same code and call it everywhere...is this buying us anything?

Reply:
It is mainly to separate concerns. Another alternative would be to have the constructor accepting the `scene_dataset_batch: Dict[int, EgoDataset]` as I mentioned, and a static factory method that would be called something like: `SimulationDataset.from_dataset_indices(dataset, indices, ...)` that will return an instance of the `SimulationDataset` after filtering (that static method will do the filtering and call the constructor). This is mainly to separate how the filtering is done, because otherwise there is no way to change it without overriding the constructor, which violates the open/closed principle that classes should be open for extension but closed for changes. If that requires too many changes, then another alternative also would be to just add the filtering code into a separate method, so that can be changed easily later without having to change the class itself.

Reply:
ok, I'll add the static factory method :) thanks for the clarification